format-avro/src/main/java/io/cdap/plugin/format/avro/AvroToStructuredTransformer.java

```diff
@@ -91,9 +95,76 @@ public Schema convertSchema(org.apache.avro.Schema schema) throws IOException {
     if (schemaCache.containsKey(hashCode)) {
       structuredSchema = schemaCache.get(hashCode);
     } else {
-      structuredSchema = Schema.parseJson(schema.toString());
+      String strSchema = schema.toString();
+      String jsonSchema = preprocessSchema(new Gson().fromJson(strSchema, JsonObject.class)).toString();
+      structuredSchema = Schema.parseJson(jsonSchema);
       schemaCache.put(hashCode, structuredSchema);
     }
     return structuredSchema;
   }
+
+  private JsonObject preprocessSchema(JsonObject schema) {
+
+    if (!schema.has("type")) {
+      return schema;
+    }
+
+    JsonElement type = schema.get("type");
+
+    if (type.isJsonArray()) {   // Union
+      for (JsonElement subtype : type.getAsJsonArray()) {
+        if (!subtype.isJsonPrimitive()) {
+          preprocessSchema(subtype.getAsJsonObject());
+        } else if (customTypes.containsKey(subtype.getAsString())) {
+          schema.remove("type");
+          schema.add("type", customTypes.get(subtype.getAsString()));
```

Comment:
Could you please explain what are you doing here (probably even with a comment). I see that for a union you are replacing the whole thing with just a subtype, e.g. `{type:[A,B,C]}` will be replaced with `{type:$(customTypes.get("C")}`. Why?

Reply:
That's an error - if `{type:[A,B,C]}` is found out of which `C` is a custom type, then only `C` should be replaced. Making that change now.